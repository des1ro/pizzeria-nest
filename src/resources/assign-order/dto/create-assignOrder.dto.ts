import { IsString } from 'class-validator';

export class CreateAssignOrderDto {
  @IsString()
  employeeId: string;
}
