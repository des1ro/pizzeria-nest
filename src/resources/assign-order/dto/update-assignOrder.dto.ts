import { IsEnum } from 'class-validator';
import { TaskStatus } from '../enum/taskStatus.enum';

export class UpdateAssignOrderDto {
  @IsEnum(TaskStatus)
  status: TaskStatus;
}
