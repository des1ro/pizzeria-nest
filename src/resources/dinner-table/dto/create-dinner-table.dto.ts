import { IsBoolean, IsNumber, IsOptional, IsString } from 'class-validator';

export class CreateDinnerTableDto {
  @IsString()
  name: string;
  @IsNumber()
  seats: number;
  @IsOptional()
  @IsBoolean()
  active?: boolean = true;
}
