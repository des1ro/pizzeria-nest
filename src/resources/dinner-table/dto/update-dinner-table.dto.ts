import { IsBoolean, IsNumber, IsOptional } from 'class-validator';

export class UpdateDinnerTableDto {
  @IsOptional()
  @IsNumber()
  seats?: number;
  @IsOptional()
  @IsBoolean()
  active?: boolean;
}
