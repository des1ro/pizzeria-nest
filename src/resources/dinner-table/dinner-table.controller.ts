import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Patch,
  Post,
} from '@nestjs/common';
import { DinnerTableService } from './dinner-table.service';
import { CreateDinnerTableDto } from './dto/create-dinner-table.dto';
import { UpdateDinnerTableDto } from './dto/update-dinner-table.dto';
import { DinnerTable } from './entities/dinner-table.entity';
import { RemoveResponse } from '../../shared/interfaces/removeResponse.interface';

@Controller('dinner-table')
export class DinnerTableController {
  constructor(private readonly dinnerTableService: DinnerTableService) {}

  @Post()
  async create(
    @Body() createDinnerTableDto: CreateDinnerTableDto,
  ): Promise<DinnerTable> {
    return await this.dinnerTableService.create(createDinnerTableDto);
  }

  @Get()
  async findAll(): Promise<DinnerTable[]> {
    return await this.dinnerTableService.findAll();
  }

  @Get(':id')
  async findOne(@Param('id') id: string): Promise<DinnerTable> {
    return await this.dinnerTableService.findOne(id);
  }

  @Patch(':id')
  async update(
    @Param('id') id: string,
    @Body() updateDinnerTableDto: UpdateDinnerTableDto,
  ): Promise<DinnerTable> {
    return await this.dinnerTableService.update(id, updateDinnerTableDto);
  }

  @Delete(':id')
  async remove(@Param('id') id: string): Promise<RemoveResponse> {
    return await this.dinnerTableService.remove(id);
  }
}
