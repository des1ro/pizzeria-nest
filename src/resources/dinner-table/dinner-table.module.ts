import { Module } from '@nestjs/common';
import { DinnerTableService } from './dinner-table.service';
import { DinnerTableController } from './dinner-table.controller';
import { TypeOrmModule } from '@nestjs/typeorm';
import { DinnerTable } from './entities/dinner-table.entity';
import { DinnerTableRepository } from './dinner-table.repo';

@Module({
  imports: [TypeOrmModule.forFeature([DinnerTable])],
  controllers: [DinnerTableController],
  providers: [DinnerTableService, DinnerTableRepository],
  exports: [DinnerTableService],
})
export class DinnerTableModule {}
