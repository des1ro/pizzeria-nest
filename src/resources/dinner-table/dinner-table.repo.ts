import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { MoreThanOrEqual, Repository } from 'typeorm';
import { DinnerTable } from './entities/dinner-table.entity';

@Injectable()
export class DinnerTableRepository {
  constructor(
    @InjectRepository(DinnerTable)
    private readonly dinnerTableRepository: Repository<DinnerTable>,
  ) {}

  async save(dinnerTable: DinnerTable) {
    return await this.dinnerTableRepository.save(dinnerTable);
  }
  async find(id: string) {
    return await this.dinnerTableRepository.findOneBy({ id });
  }
  async findManyBySeats(seats: number): Promise<DinnerTable[]> {
    return await this.dinnerTableRepository.find({
      where: {
        seats: MoreThanOrEqual(seats),
        active: true,
      },
    });
  }
  async findAll() {
    return await this.dinnerTableRepository.find();
  }
  async remove(dinnerTable: DinnerTable) {
    return this.dinnerTableRepository.remove(dinnerTable);
  }
}
