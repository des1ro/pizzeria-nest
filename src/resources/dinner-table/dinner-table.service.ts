import { Injectable, NotFoundException } from '@nestjs/common';
import { CreateDinnerTableDto } from './dto/create-dinner-table.dto';
import { UpdateDinnerTableDto } from './dto/update-dinner-table.dto';
import { DinnerTable } from './entities/dinner-table.entity';
import { DinnerTableRepository } from './dinner-table.repo';
import { RemoveResponse } from '../../shared/interfaces/removeResponse.interface';

@Injectable()
export class DinnerTableService {
  constructor(private readonly repository: DinnerTableRepository) {}
  create(createDinnerTableDto: CreateDinnerTableDto) {
    const dinnerTable = new DinnerTable(createDinnerTableDto);
    return this.repository.save(dinnerTable);
  }
  async update(
    id: string,
    updateDinnerTableDto: UpdateDinnerTableDto,
  ): Promise<DinnerTable> {
    const dinnerTable = await this.findOne(id);
    Object.assign(dinnerTable, updateDinnerTableDto);
    return this.repository.save(dinnerTable);
  }

  async findAll(): Promise<DinnerTable[]> {
    const dinnerTables = await this.repository.findAll();
    if (dinnerTables.length === 0) {
      throw new NotFoundException(`Dinner Tables not found`);
    }
    return dinnerTables;
  }

  async findOne(id: string): Promise<DinnerTable> {
    const dinnerTable = await this.repository.find(id);
    if (!dinnerTable) {
      throw new NotFoundException(`Dinner Table with ID ${id} not found`);
    }
    return dinnerTable;
  }

  async remove(id: string): Promise<RemoveResponse> {
    const dinnerTable = await this.findOne(id);
    await this.repository.remove(dinnerTable);
    return {
      success: true,
      message: `Dinner Table with ID ${id} removed successfuly`,
    };
  }

  async findBySeats(seats: number): Promise<DinnerTable> {
    const dinnerTables = await this.repository.findManyBySeats(seats);
    if (dinnerTables.length === 0) {
      throw new NotFoundException(
        `Not found any free Dinner Table with ${seats} seats`,
      );
    }
    let dinnerTable: DinnerTable;
    for (const table of dinnerTables) {
      if (table.seats === seats) {
        dinnerTable = table;
        break;
      }
      if (!dinnerTable || dinnerTable.seats > table.seats) {
        dinnerTable = table;
      }
    }
    return dinnerTable;
  }
}
