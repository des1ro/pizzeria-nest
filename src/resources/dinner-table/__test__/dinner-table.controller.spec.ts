import { Test, TestingModule } from '@nestjs/testing';
import { DinnerTableController } from '../dinner-table.controller';
import { DinnerTableService } from '../dinner-table.service';
import { CreateDinnerTableDto } from '../dto/create-dinner-table.dto';
import { UpdateDinnerTableDto } from '../dto/update-dinner-table.dto';

describe('DinnerTableController test suite', () => {
  let objectUnderTest: DinnerTableController;
  let mockedCreateDinnerTableDtoOne: CreateDinnerTableDto;
  let mockedCreateDinnerTableDtoTwo: CreateDinnerTableDto;
  const testIdOne = 'test uuid one';
  const testIdTwo = 'test uuid two';
  const mockedService = {
    create: jest.fn((dto) => {
      return { id: testIdOne, ...dto };
    }),
    findAll: jest.fn(() => {
      return [
        {
          id: testIdOne,
          ...mockedCreateDinnerTableDtoOne,
        },
        { id: testIdTwo, ...mockedCreateDinnerTableDtoTwo },
      ];
    }),

    findOne: jest.fn((id) => {
      return { id: id, ...mockedCreateDinnerTableDtoOne };
    }),
    update: jest.fn((id) => {
      return { id: id, ...mockedCreateDinnerTableDtoOne };
    }),
    remove: jest.fn(() => {
      return { success: true, message: 'test message' };
    }),
  };
  beforeAll(() => {
    mockedCreateDinnerTableDtoOne = {
      name: 'test dinner-table',
      seats: 1,
      active: true,
    };
    mockedCreateDinnerTableDtoTwo = {
      name: 'test dinner-table',
      seats: 1,
      active: true,
    };
  });
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [DinnerTableController],
      providers: [DinnerTableService],
    })
      .overrideProvider(DinnerTableService)
      .useValue(mockedService)
      .compile();

    objectUnderTest = module.get<DinnerTableController>(DinnerTableController);
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(objectUnderTest).toBeDefined();
  });
  describe('create dinner-table test suite', () => {
    it('should create dinner-table', async () => {
      expect(
        await objectUnderTest.create(mockedCreateDinnerTableDtoOne),
      ).toEqual({
        id: expect.any(String),
        ...mockedCreateDinnerTableDtoOne,
      });
      expect(mockedService.create).toHaveBeenCalledTimes(1);
    });
    it("should set to active if it's not provided", async () => {
      //Given
      mockedCreateDinnerTableDtoOne.active = undefined;
      //When
      const expectedResult = mockedCreateDinnerTableDtoOne;
      expectedResult.active = true;
      //Then
      expect(
        await objectUnderTest.create(mockedCreateDinnerTableDtoOne),
      ).toEqual({
        id: expect.any(String),
        ...expectedResult,
      });

      expect(objectUnderTest).toBeDefined();
      expect(mockedService.create).toHaveBeenCalledTimes(1);
    });
  });

  it('should find all dinner-tables', async () => {
    //When
    const result = await objectUnderTest.findAll();
    //Then
    expect(result).toMatchSnapshot('dinner tables');
    expect(mockedService.findAll).toHaveBeenCalledTimes(1);
  });
  it('should findOne dinner-table', async () => {
    //When
    const result = await objectUnderTest.findOne(testIdOne);
    //Then
    expect(result.id).toEqual(testIdOne);
    expect(mockedService.findOne).toHaveBeenCalledTimes(1);
  });
  it('should update dinner-table', async () => {
    //Given
    const mockedUpdateDto = new UpdateDinnerTableDto();
    //When
    const result = await objectUnderTest.update(testIdOne, mockedUpdateDto);
    //Then
    expect(result.id).toEqual(testIdOne);
    expect(mockedService.update).toHaveBeenCalledTimes(1);
  });
  it('should remove dinner-table and return response', async () => {
    //Then
    expect(await objectUnderTest.remove(testIdOne)).toEqual({
      success: true,
      message: expect.any(String),
    });
  });
});
