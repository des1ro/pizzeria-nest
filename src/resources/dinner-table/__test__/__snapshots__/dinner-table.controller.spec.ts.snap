// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DinnerTableController test suite should find all dinner-tables: dinner tables 1`] = `
[
  {
    "active": true,
    "id": "test uuid one",
    "name": "test dinner-table",
    "seats": 1,
  },
  {
    "active": true,
    "id": "test uuid two",
    "name": "test dinner-table",
    "seats": 1,
  },
]
`;
